### PR TITLE
[ROCm] [Model] Enable TML inkling

### PR DESCRIPTION
**Pull request by tjtanaa**
## Purpose

This is the first enablement PR for inkling model on ROCm. This PR also targets the more efficient variant of the model which is mxfp4. There are possible path for loading and use nvfp4 weight, however, we could only support it through inefficient emulation code path. 

The weight is generated and distributed by https://huggingface.co/lightseekorg/Inkling-MXFP4 instead. 

There is a continuous effort to try to enable to BF16 weight and NVFP4 weights.

## Test Result

```
#!/bin/bash
set -euo pipefail

MODEL=lightseekorg/Inkling-MXFP4

export VLLM_USE_V2_MODEL_RUNNER=1
export VLLM_ROCM_USE_AITER=1
export INKLING_MULTIMEM_AR=0

"vllm serve "$MODEL_SNAPSHOT" \
      --served-model-name "$MODEL" \
      --tokenizer-mode inkling \
      --reasoning-parser inkling \
      --tool-call-parser inkling \
      --enable-auto-tool-choice \
      --tensor-parallel-size 8 \
      --max-model-len 65536 \
      --gpu-memory-utilization 0.80 \
      --trust-remote-code
```

GSM8K 

| Result | Page 16 | Page 128 |
|---|---:|---:|
| Flexible exact match | 95.22% | **95.83%** |
| Strict exact match | 94.01% | **94.92%** |

### BFCL Live AST

Command settings:

- 1,351 calls
- BFCL evaluation version `2025.12.17`
- concurrency 128
- reasoning effort 0.99
- temperature 0
- maximum generation length 4,096

BFCL :        78.83%  (this PR) vs vLLM Blog (78.61%)

| Result | Page 16 | Page 128 | Page 128 vs 16 | vLLM Blog target |
|---|---:|---:|---:|---:|
| Exact calls | 78.39% (1,059/1,351) | **78.76% (1,064/1,351)** | +0.37 pp | 78.61% |
| All-Live macro | 78.84% | **81.09%** | +2.25 pp | 75.86% |
| Runtime | 116.8s | **111.5s** | 5.2s faster | — |
| Within expected exact range | Yes | Yes | — | 76.42–80.80% |

Also support the kernel from tokenspeed

| Metric | Page 16 | Page 128 | Page-128 change |
|---|---:|---:|---:|
| Benchmark duration | 282.84s | **219.28s** | -22.47% |
| Request throughput | 1.131 req/s | **1.459 req/s** | +28.99% |
| Output throughput | 1,158.54 tok/s | **1,494.37 tok/s** | +28.99% |
| Total token throughput | 2,334.06 tok/s | **3,010.63 tok/s** | +28.99% |
| Peak output throughput | 1,374 tok/s | **1,663 tok/s** | +21.03% |
| Mean TTFT | 579.68 ms | **574.83 ms** | -0.84% |
| P99 TTFT | 1,075.49 ms | **1,064.53 ms** | -1.02% |
| Mean TPOT | 27.08 ms | **20.87 ms** | -22.93% |
| P99 TPOT | 27.46 ms | **21.27 ms** | -22.54% |
| Mean ITL | 28.00 ms | **21.51 ms** | -23.19% |
| P99 ITL | 53.31 ms | **40.74 ms** | -23.57% |
| Mean end-to-end latency | 28.28s | **21.92s** | -22.47% |
| P99 end-to-end latency | 28.55s | **22.20s** | -22.23% |

Page 128 is approximately **1.29x faster** on throughput and **1.30x faster** on per-output-token latency. TTFT is effectively unchanged, which is consistent with the improvement primarily coming from decode.

### MTP-1

With  `--speculative-config '{"method":"mtp","num_speculative_tokens":1}'`

| Page size | GSM8K flexible | GSM8K strict | Accepted / drafted | Draft acceptance | Mean acceptance length |
|---:|---:|---:|---:|---:|---:|
| 16 | 95.38% | 93.71% | 153,444 / 185,058 | 82.92% | 1.83 |
| 128 | 95.53% | 93.71% | 150,376 / 181,449 | 82.88% | 1.83 |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: tests/models/inkling/rocm/test_mtp_input_fusion.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+"""
+
+from typing import cast
+
+import pytest
+import torch
+
+from vllm.platforms import current_platform
+
+if not current_platform.is_rocm():
```
Comment (AndreasKaratzas) [inline tests/models/inkling/rocm/test_mtp_input_fusion.py]: If these tests are under rocm dir, shouldnt we just put a conftest to skip in all other platforms?

File: tests/models/inkling/rocm/test_rel_attention.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+    cache_seqlens = torch.tensor(kv_lens, device=device, dtype=torch.int32)
+    cu_seqlens_q = torch.arange(
+        batch_size + 1,
+        device=device,
+        dtype=torch.int32,
+    )
+    window_size = (-1, -1) if window_left is None else (window_left, 0)
+    max_kv_len = 8192 if page_size == 16 else max(kv_lens)
+
+    actual = inkling_fa4_rel_attention(
```
Comment (AndreasKaratzas) [inline tests/models/inkling/rocm/test_rel_attention.py:285]: Do we need to add a skip on any ROCm platform that is not gfx950?

File: vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py [hunk 3]
```diff
@@ -0,0 +1,1037 @@
...
+
+    config = get_config(
+        q=q,
+        k_cache=k_cache,
+        max_seqlen_k=max_seqlen_k,
+        window_left=window_left,
+        softmax_scale=softmax_scale,
+    )
+
+    is_fp8 = q.dtype in (torch.float8_e4m3fn, torch.float8_e5m2)
```
Comment (AndreasKaratzas) [inline vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py:945]: nit: would it be better to use q.dtype is current_platform.fp8_dtype() ?

File: vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py [hunk 4]
```diff
@@ -0,0 +1,1037 @@
...
+            f"64, 128, or 256; got {k_cache.shape[1]}"
+        )
+    output = (
+        torch.empty(q.shape, device=q.device, dtype=out_dtype) if out is None else out
+    )
+
+    # Always use split-k for full attention and for the small-batch sliding
+    # decode path. Sliding uses a fixed 8 splits, one page per split for the
+    # TP-4 local-attention shape with 512-token window and 64-token pages.
+    mid_o = torch.empty(
```
Comment (AndreasKaratzas) [inline vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py:959]: sometimes, torch.empty creates issue on rocm as opposed to torch.zeros (let's make sure that this is not such a case)

File: vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py (hunk 3 above)
Comment (tjtanaa) [reply vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py:945]: I would like to avoid this in this case as this kernel is explicitly written for gfx950 from tokenspeed repo. It is an implementation from there. I would like to get the code into this repo so that we can maintain a stable version of the code and also avoid installing studs. If we `pip install tokenspeed-kernel-amd`, it will install its own triton and gluon stud which could be conflicting with our customly shipped torch and triton version. On top of that, the optimization is very significant that the plan is to have the kernel enabled by default.

File: vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py (hunk 4 above)
Comment (tjtanaa) [reply vllm/models/inkling/amd/ops/gluon/rel_mha_decode_gfx950.py:959]: The kernel has been validated very extensively on tokenspeed repo as shown in https://lightseek.org/blog/tokenspeed-inkling.html when they are supporting inkling on ROCm https://github.com/lightseekorg/tokenspeed/pull/689/changes 

so, this kernel should be good. I can come back to add more tests in vllm if there is a concern.

File: tests/models/inkling/rocm/test_mtp_input_fusion.py (hunk 1 above)
Comment (tjtanaa) [reply tests/models/inkling/rocm/test_mtp_input_fusion.py]: great suggestion. I added a conftest

File: tests/models/inkling/rocm/test_rel_attention.py (hunk 2 above)
Comment (tjtanaa) [reply tests/models/inkling/rocm/test_rel_attention.py:285]: done 👍

File: tests/models/inkling/rocm/conftest.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+        if not on_gfx950():
+            pytest.skip("requires CDNA4")
+        return
+
+    rocm_test_dir = Path(__file__).parent
+    skip_rocm = pytest.mark.skip(reason="requires ROCm")
+    for item in items:
+        if item.path.is_relative_to(rocm_test_dir):
+            item.add_marker(skip_rocm)
```
Comment (AndreasKaratzas) [inline tests/models/inkling/rocm/conftest.py]: Is this necessary? Can't we just add a fixture in a separate function to just pytest skip if not is rocm? (maybe it's not possible, just asking 😅 )

Comment (tjtanaa) [reply tests/models/inkling/rocm/conftest.py]: How about this current implementation.

**Review by AndreasKaratzas — APPROVED** [review]
LGTM

File: tests/models/inkling/rocm/test_rocm_mtp_input_fusion.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tjtanaa) [inline tests/models/inkling/rocm/test_rocm_mtp_input_fusion.py:1]: Cause: pytest_s default import mode treated both files as the top-level module test_mtp_input_fusion:

  - `tests/models/inkling/test_mtp_input_fusion.py`
  - `tests/models/inkling/rocm/test_mtp_input_fusion.py`


So I have renamed `tests/models/inkling/rocm/test_mtp_input_fusion.py` to `tests/models/inkling/rocm/test_rocm_mtp_input_fusion.py`
